File: web/src/lib/sitemap.ts
// Shared builders for the sitemap index and its sub-sitemaps. The catalogue is
// far larger than the 50,000-URL / 50 MB sitemap-protocol limit, so /sitemap.xml
// is an index that points at chunked sub-sitemaps (static pages, job chunks,
// company chunks). Each chunk is one keyset page fetched by cursor.

import { collectionSlugs } from './collections';
import { contributorGroups } from './contributors';
import type { ContributorsSnapshot } from './contributors';
import contributorsSnapshot from './data/contributors.json';

// Must equal the backend's companySitemapChunk — the page size the offsets are
// computed with — so each sub-sitemap holds exactly one index page. Well under the
// protocol's 50,000-URL cap, and left at the size the Postgres-backed version used
// so the existing sub-sitemap URLs keep pointing at the same tiling.
export const SITEMAP_CHUNK = 10000;

// Must equal the backend's jobSitemapChunk. Job chunks are pages of the search
// index, not of the jobs table: the index holds every posting the site's own search
// can find, and it can address any offset in it directly — which is what let the
// sitemap stop asking Postgres to number 3.4M rows on every render.
//
// Note the chunks page a NARROWED view of that index, not all of it: the backend's
// jobSitemapFilter admits only tech postings that are not flagged as perpetual
// listings, and the count these offsets are computed from is filtered the same way.
// So the number of job chunks tracks the tech slice (~669k, 68 chunks) rather than
// the whole index (~2.0M) — a sub-sitemap URL for a retired offset answers with an
// empty urlset, which is what the route already promises for any offset past the end.
//
// 10k rather than 25k because the deepest 25k page measured 8s against this route's
// 10s fetch timeout under load — see the backend constant for the trade.
export const JOB_SITEMAP_CHUNK = 10000;

// How many chunks the FRESHEST-first job sub-sitemap spans. Must not exceed the
// backend's freshSitemapMaxOffset tiling (`3 * jobSitemapChunk`, i.e. four pages):
// past that bound the API answers an empty page on purpose, so a fifth entry here
// would put an empty file in the index rather than extending the window.
//
// Four chunks is 40,000 URLs, about 3.5 days of this catalogue's intake, so a crawler
// that skips two days still misses nothing. The bound exists because this file is
// SORTED, and unlike the paged chunks a sorted read gets more expensive with depth —
// see the backend constant for the measurements.
//
// Changing the count means moving BOTH this and freshSitemapMaxOffset. Moving only the
// Go side leaves an entry here pointing at a file the API deliberately answers empty,
// which looks exactly like a working sitemap — the same two-sided coupling
// JOB_SITEMAP_CHUNK already has with jobSitemapChunk.
export const FRESH_SITEMAP_CHUNKS = 4;

/** The offset opening each chunk of the freshest-first job sub-sitemap.
 *
 *  Derived from the two constants rather than written out, so the tiling cannot drift
 *  from the page size the API serves: a hard-coded list would survive a change to
 *  JOB_SITEMAP_CHUNK and then address offsets that fall between pages. */
export function freshJobSitemapOffsets(): number[] {
  return Array.from({ length: FRESH_SITEMAP_CHUNKS }, (_, i) => i * JOB_SITEMAP_CHUNK);
}

/** Every `<loc>` of the sitemap index, in the order a crawler reads them.
 *
 *  A pure function rather than inline in the route because the ORDER is a requirement
 *  (web-ssr-seo: the freshest-first sub-sitemaps precede the paged ones) and an order
 *  built inside a route handler has nothing to assert on but a rendered XML string.
 *  The route still owns the two boundary reads; this owns what to do with them. */
export function sitemapIndexLocs(params: {
  origin: string;
  roleCategorySlugs: string[];
  jobOffsets: number[];
  companyOffsets: number[];
}): string[] {
  const { origin, roleCategorySlugs, jobOffsets, companyOffsets } = params;
  const locs = [`${origin}/sitemap-pages.xml`, `${origin}/sitemap-insights.xml`];
  // The skills glossary is one file rather than a shard per letter: it is under a
  // thousand URLs and enumerating it reads nothing, so there is nothing to shard away
  // from — unlike the role landings below, where each shard pays its own facet call.
  locs.push(`${origin}/sitemap-skills.xml`);
  // One role sub-sitemap per category. The category list is a compile-time constant,
  // so naming all of them costs no read here — each shard pays its own single facet
  // call when a crawler actually follows it.
  for (const slug of roleCategorySlugs) {
    locs.push(`${origin}/sitemap-roles.xml?category=${slug}`);
  }
  // The freshest-first job chunks come BEFORE the paged ones. Position in a sitemap
  // index binds no crawler, but it is the only signal the format offers and it costs
  // nothing — and the paged chunks that follow are in arbitrary order, so without this
  // a crawler has no way at all to find what is new.
  for (const offset of freshJobSitemapOffsets()) {
    locs.push(`${origin}/sitemap-jobs-fresh.xml?offset=${offset}`);
  }
  // Both cursor lists already include the opening 0, so they are listed as they come.
  for (const offset of jobOffsets) {
    locs.push(`${origin}/sitemap-jobs.xml?offset=${offset}`);
  }
  for (const offset of companyOffsets) {
    locs.push(`${origin}/sitemap-companies.xml?offset=${offset}`);
  }
  return locs;
}

/** The site's static, always-present pages (relative paths). */
export const STATIC_PATHS = [
  '/',
  // The job feed. It used to BE `/`, which is now the landing page — so this entry is
  // not a second copy of the homepage but the URL every filtered and paginated feed
  // link resolves to, and the one page here that carries the catalogue itself.
  '/jobs',
  '/about',
  '/companies',
  '/collections',
  // The public Talent Network catalogue. Indexable and linked from the header menu
  // and the footer, but it was in neither the sitemap nor anything a crawler reads,
  // so nothing told a search engine it exists. The per-candidate cards under it are
  // NOT here: a card is anonymous by design and 404s until its owner's CV has been
  // read, so offering crawlers a path to one is offering a path to a page we cannot
  // promise resolves.
  '/talent',
  '/for-companies',
  '/recruiters',
  '/features/extension',
  '/features/inbox',
  '/features/referrals',
  '/features/tailor',
  '/features/tracking',
  '/features/notifications',
  '/features/ghost-jobs',
  '/features/advanced-search',
  '/how-it-works',
  // The discussions feed. The individual threads are NOT here: they live under their
  // subject's route, and a sub-sitemap for the handful that exist would be
  // infrastructure ahead of need.
  '/discussions',
  // The data and API surfaces. Indexable pages that carry the site's most citable
  // material — live catalogue figures (/open), market rollups (/trends), the API
  // reference — so they belong in the sitemap even though nothing links to some of
  // them from the feed.
  '/open',
  // The market rollups. /analytics was reachable from the header menu and from
  // nowhere a crawler reads, the same gap /talent had.
  '/analytics',
  // Where the catalogue comes from: every source, what each carries, when it was last
  // read. Sits with /open rather than with /status because it is a citable statement
  // about the catalogue, not an operational dashboard.
  '/sources',
  '/trends',
  '/docs/api',
  '/agents',
  '/cli',
  '/chatgpt',
  // The public CV roast: an account-free landing page nothing in the feed links to,
  // built to be found by search rather than clicked to from elsewhere on the site.
  '/roast',
  '/contribute',
  // The contributor showcase. The per-person profiles are NOT here — they come from
  // the committed snapshot via contributorPaths(), the same way collections do.
  '/contributors',
  '/status',
  // Named as the App Store listing's Support URL, so it must resolve and stay
  // indexable — a 404 there is a rejection.
  '/support',
  '/privacy',
  // The role×country hub. Its category and pair pages are gated on live counts, so
  // they ride in their own per-category sub-sitemaps rather than here.
  '/roles',
];

/** The curated collection landing pages (`/collections/:slug`), one per collection.
 *  A small, fixed set, so they ride in the static-pages sub-sitemap alongside
 *  STATIC_PATHS rather than needing their own chunked file. */
export function collectionPaths(): string[] {
  return collectionSlugs().map((slug) => `/collections/${slug}`);
}

/** The per-contributor profile pages (`/contributors/:login`), one per person the
 *  showcase actually lists.
 *
 *  Derived through the same rules the showcase renders, not from the raw snapshot: a
 *  bot's profile 404s, so offering crawlers a path to one would be pointing them at a
 *  page we know does not exist. Each profile is its own indexable page because it is
 *  the page a person shares and is found by. */
export function contributorPaths(): string[] {
  const { maintainers, contributors } = contributorGroups(contributorsSnapshot as ContributorsSnapshot);

  return [...maintainers, ...contributors].map((entry) => `/contributors/${entry.login}`);
}

/** Sitemap entries for the blog: the index (`/blog`) plus one per published post,
 *  each dated with the post's own publish date. Takes the posts (from
 *  `listPosts()`) rather than reading them itself, so it stays pure/testable — the
 *  glob-backed loader is called by the route.
 *
 *  The dates matter: without a `lastmod` a crawler has nothing to tell an edited
 *  or newly published post from one it already has, and re-reads the whole set on
 *  its own schedule. The index carries the newest post's date, since that is
 *  exactly when its content last changed. */
export function blogPaths(posts: { slug: string; date: string }[]): PathEntry[] {
  const newest = posts.map((post) => post.date).sort().at(-1);
  return [
    { path: '/blog', lastmod: newest },
    ...posts.map((post) => ({ path: `/blog/${post.slug}`, lastmod: post.date })),
  ];
}

/** Sitemap paths for the insights pages: the hub plus salary/skills/roles for each
 *  covered category. Takes the already-gated category tokens (from
 *  `coveredCategories`) so it stays pure — a thin category is never listed. */
export function insightsPaths(categories: string[], roleLeaves: [string, string][] = []): string[] {
  const paths = ['/insights'];
  for (const c of categories) {
    paths.push(`/insights/salary/${c}`, `/insights/skills/${c}`, `/insights/roles/${c}`);
  }
  // The per-role leaves, passed already gated (from `coveredRoles`) so a pair the
  // route would 404 is never listed — the same rule read off the same numbers that
  // `roleLandingPaths` follows for the jobs landings.
  for (const [category, seniority] of roleLeaves) {
    paths.push(`/insights/roles/${category}/${seniority}`);
  }
  return paths;
}

/** Sitemap paths for one category's landings: its country table plus a page per
 *  country that clears the gate. Takes the already-gated countries (from
 *  `publishedCountries`) so it stays pure — a thin pair is never listed, matching
 *  what the route serves for one (404).
 *
 *  Split per category on purpose: the whole product is ~2,200 URLs, which fits one
 *  file, but building it in one would cost one facet call per category in a single
 *  request. One sub-sitemap per category costs one call each, and the index can name
 *  all 37 without reading anything. */
export function roleLandingPaths(categorySlug: string, countrySlugs: string[]): string[] {
  return [
    `/roles/${categorySlug}`,
    ...countrySlugs.map((country) => `/roles/${categorySlug}/${country}`),
  ];
}

/** Sitemap paths for the skill glossary: the index plus one page per skill that has a
 *  description.
 *
 *  It takes the described slugs rather than the whole canonical vocabulary because the
 *  route 404s on a skill with no entry — listing all of them would point a crawler at
 *  pages that do not exist, and a sitemap full of 404s is worse than a short one. The
 *  caller reads the set from the same catalog the route does, so the two cannot drift.
 *
 *  One file, unlike the role landings: the whole glossary is under a thousand URLs and
 *  costs no read at all to enumerate, so there is nothing to shard away from. */
export function skillGlossaryPaths(describedSlugs: readonly string[]): string[] {
  return ['/skills', ...describedSlugs.map((slug) => `/skills/${slug}`)];
}

/** A sitemap path with the date its content last changed, before an origin is
 *  prefixed. `lastmod` is undefined when there is no honest date to state — a
 *  guessed one is worse than none, since a crawler that learns the dates are
 *  noise stops using them. */
export interface PathEntry {
  path: string;
  lastmod?: string;
}

export interface UrlEntry {
  loc: string;
  lastmod?: string;
}

function escapeXml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function urlTag({ loc, lastmod }: UrlEntry): string {
  const mod = lastmod ? `\n    <lastmod>${escapeXml(lastmod)}</lastmod>` : '';
  return `  <url>\n    <loc>${escapeXml(loc)}</loc>${mod}\n  </url>`;
}

/** A `<urlset>` sub-sitemap document from page URLs. */
export function urlsetXml(entries: UrlEntry[]): string {
  return `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${entries.map(urlTag).join('\n')}
</urlset>
`;
}

/** A `<sitemapindex>` document referencing sub-sitemap URLs. */
export function sitemapIndexXml(locs: string[]): string {
  const items = locs.map((loc) => `  <sitemap>\n    <loc>${escapeXml(loc)}</loc>\n  </sitemap>`).join('\n');
  return `<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
${items}
</sitemapindex>
`;
}

/** Wrap sitemap XML in a cached response (crawlers/CDN don't re-run the paging). */
export function xmlResponse(body: string): Response {
  return new Response(body, {
    headers: {
      'content-type': 'application/xml; charset=utf-8',
      'cache-control': 'public, max-age=3600',
    },
  });
}

File: internal/handler/sitemap.go
package handler

import (
	"context"
	"time"

	"github.com/gofiber/fiber/v2"

	"github.com/strelov1/freehire/internal/db"
	"github.com/strelov1/freehire/internal/search"
)

// sitemapLister is the search-index read the job sitemap pages over: one
// offset-addressed page of indexed jobs plus the index's total document count.
// Narrow on purpose — the sitemap needs a slug and a lastmod, nothing the search
// handler's `searcher` offers.
type sitemapLister interface {
	ListSitemapPage(ctx context.Context, offset, limit int) ([]search.SitemapDocument, int64, error)
	CountSitemapDocuments(ctx context.Context) (int64, error)
}

// sitemapHandlers serves the XML sitemaps for jobs and companies. The company
// sitemap literals are registered before the /companies/:slug param route (in
// Register) so they are not read as slugs; likewise /jobs/sitemap precedes
// /jobs/:slug.
//
// Jobs and companies are paged differently because their sources are: companies
// are keyset-paged out of Postgres by slug, while jobs are offset-paged out of the
// Meilisearch index (see jobs below).
type sitemapHandlers struct {
	queries *db.Queries
	jobs    sitemapLister
}

func newSitemapHandlers(queries *db.Queries, jobs sitemapLister) *sitemapHandlers {
	return &sitemapHandlers{queries: queries, jobs: jobs}
}

func (h *sitemapHandlers) register(api fiber.Router) {
	api.Get("/jobs/sitemap", h.JobSitemap)
	api.Get("/jobs/sitemap/boundaries", h.JobSitemapBoundaries)
	api.Get("/companies/sitemap", h.CompanySitemap)
	api.Get("/companies/sitemap/boundaries", h.CompanySitemapBoundaries)
}

// sitemapMaxURLs is the sitemap-protocol per-file cap — the hard ceiling an
// untrusted ?limit= / ?chunk= is clamped to, so a served chunk can never exceed the
// protocol limit however it is asked for.
const sitemapMaxURLs = 50000

// companySitemapChunk is how many companies one sub-sitemap holds: the default for
// both ?limit= and ?chunk=, and the value web/src/lib/sitemap.ts SITEMAP_CHUNK must
// equal so each boundary cursor opens exactly one file's worth.
//
// Deliberately far below the protocol cap. These reads compete with the ingest for
// the buffer cache and their latency swings with it by more than an order of
// magnitude: on prod (2026-07-29) the same 50k-row chunk measured 0.9s warm and past
// 60s — an nginx 504 — while an ingest run was evicting the cache. A tenth of the
// rows keeps the slow end inside the proxy timeout; the cost is ~21 files instead of
// ~5, which a sitemap index carries for free (its own cap is 50k entries).
const companySitemapChunk = 10000

// jobSitemapChunk is how many jobs one sub-sitemap holds: the default for both
// ?limit= and ?chunk=, and the value web/src/lib/sitemap.ts JOB_SITEMAP_CHUNK must
// equal so each offset in the index opens exactly one file's worth.
//
// Kept below the protocol's 50k cap with room to spare: a 25k page measured ~2.5s
// against the prod index, and doubling it would put a single sub-sitemap's render
// near the SSR timeout for no gain — a sitemap index carries the extra files for
// free (its own cap is 50k entries).
const jobSitemapChunk = 25000

// sitemapEntry is the slim wire shape a sitemap URL needs — the public slug and a
// lastmod. Nothing wider (no full job row) crosses the wire. updated_at is NOT NULL
// on both tables, so it is always a real instant.
type sitemapEntry struct {
	Slug      string    `json:"slug"`
	UpdatedAt time.Time `json:"updated_at"`
}

// sitemapLimit clamps ?limit= to [1, sitemapMaxURLs], defaulting to `fallback`.
func sitemapLimit(c *fiber.Ctx, fallback int) int32 {
	return int32(min(max(c.QueryInt("limit", fallback), 1), sitemapMaxURLs))
}

// sitemapChunk clamps ?chunk= to [1, sitemapMaxURLs], defaulting to `fallback`.
func sitemapChunk(c *fiber.Ctx, fallback int) int64 {
	return int64(min(max(c.QueryInt("chunk", fallback), 1), sitemapMaxURLs))
}

// JobSitemap serves one page of job sitemap entries at ?offset=<n>, read from the
// search index (see search.Client.ListSitemapPage for why the index and not the
// jobs table, and why an offset and not a keyset cursor).
//
// Reads its page through the shared pageParamsBounded rather than the sitemapLimit
// helper below, because it is the one sitemap endpoint that takes an offset — and
// that parse belongs in exactly one place (see the helper, and the test that pins
// it). Absent, unparseable, or out-of-range values collapse to the first page, and
// an offset past the end is a valid empty page: a crawler holding a stale sitemap
// index must never be answered with an error.
func (h *sitemapHandlers) JobSitemap(c *fiber.Ctx) error {
	if h.jobs == nil {
		return fiber.NewError(fiber.StatusServiceUnavailable, "search is not available")
	}
	limit, offset := pageParamsBounded(c, jobSitemapChunk, sitemapMaxURLs)
	docs, _, err := h.jobs.ListSitemapPage(c.Context(), offset, limit)
	if err != nil {
		return err
	}
	entries := make([]sitemapEntry, len(docs))
	for i, d := range docs {
		entries[i] = sitemapEntry{Slug: d.Slug, UpdatedAt: d.UpdatedAt}
	}
	return c.JSON(fiber.Map{"data": entries})
}

// JobSitemapBoundaries returns the offset opening each ?chunk=<n>-sized page of the
// jobs index, for building the sitemap index — [0, chunk, 2*chunk, ...]. Same source
// as JobSitemap, so an offset always opens a page that has URLs in it.
//
// This is arithmetic over one number the engine reports for free, where the jobs
// table needed a `row_number()` walk of the whole catalogue to find the same
// boundaries — the walk that grew to 64s and started timing out (see
// search.Client.ListSitemapPage).
func (h *sitemapHandlers) JobSitemapBoundaries(c *fiber.Ctx) error {
	if h.jobs == nil {
		return fiber.NewError(fiber.StatusServiceUnavailable, "search is not available")
	}
	total, err := h.jobs.CountSitemapDocuments(c.Context())
	if err != nil {
		return err
	}
	chunk := sitemapChunk(c, jobSitemapChunk)
	offsets := make([]int64, 0, total/chunk+1)
	for off := int64(0); off < total; off += chunk {
		offsets = append(offsets, off)
	}
	return c.JSON(fiber.Map{"data": offsets})
}

// CompanySitemap serves one keyset page of company sitemap entries after ?after=<slug>,
// covering the hiring companies the /companies catalog lists (see ListCompanySitemap).
func (h *sitemapHandlers) CompanySitemap(c *fiber.Ctx) error {
	rows, err := h.queries.ListCompanySitemap(c.Context(), db.ListCompanySitemapParams{
		AfterSlug: c.Query("after"),
		BatchSize: sitemapLimit(c, companySitemapChunk),
	})
	if err != nil {
		return err
	}
	entries := make([]sitemapEntry, len(rows))
	for i, r := range rows {
		entries[i] = sitemapEntry{Slug: r.Slug, UpdatedAt: r.UpdatedAt.Time}
	}
	return c.JSON(fiber.Map{"data": entries})
}

// CompanySitemapBoundaries returns the keyset cursor (slug) ending each ?chunk=<n> of
// hiring companies, for building the sitemap index. Same scope as CompanySitemap, so a
// cursor always opens a chunk that has URLs in it.
func (h *sitemapHandlers) CompanySitemapBoundaries(c *fiber.Ctx) error {
	cursors, err := h.queries.CompanySitemapBoundaries(c.Context(), sitemapChunk(c, companySitemapChunk))
	if err != nil {
		return err
	}
	return c.JSON(fiber.Map{"data": cursors})
}
